File: scratch/migrate_articles.py
import os
import shutil
import re

articles_dir = r"C:\Users\ntana\Documents\Antigravity_docs\digital-literacy-ip\src\articles"

categories = {
    "01_p2p": ["01", "05", "06", "07", "08", "09", "10", "20", "21", "26", "27", "29", "30"],
    "02_apps": ["31", "32", "33", "34", "35", "36"],
    "03_clips": ["02", "04", "11", "12", "17"],
    "04_web": ["03", "13", "14", "15", "16", "18", "19", "22", "25", "28"]
}

# 1. Map old article names to new structure
mapping = {}
category_articles = {} # category_name -> list of new_article_names
new_titles = {} # new_article_name -> clean_title

for cat_name, old_nums in categories.items():
    cat_num = cat_name.split("_")[0]
    category_articles[cat_name] = []
    
    for i, old_num in enumerate(old_nums):
        new_num = f"{i+1:02d}"
        old_dir = f"article-{old_num}"
        new_dir = f"article-{cat_num}-{new_num}"
        mapping[old_dir] = {"cat": cat_name, "new_dir": new_dir, "file_id": f"{cat_num}-{new_num}"}
        category_articles[cat_name].append(new_dir)

# Helper to clean titles
def clean_title(title):
    # Remove existing "FILE XX: " or similar prefixes
    title = re.sub(r'^(?:FILE\s*\d+:?|FILE\s*\d+-\d+:?)\s*', '', title, flags=re.IGNORECASE)
    return title.strip()

# 2. Extract current titles before moving
old_titles = {}
for old_dir in mapping.keys():
    idx_path = os.path.join(articles_dir, old_dir, "index.html")
    if os.path.exists(idx_path):
        with open(idx_path, "r", encoding="utf-8") as f:
            content = f.read()
        m = re.search(r'<h1>(.*?)</h1>', content)
        if m:
            old_titles[old_dir] = clean_title(m.group(1))
        else:
            old_titles[old_dir] = "No Title"
    else:
        old_titles[old_dir] = "No Title"

# Map new titles
for old_dir, info in mapping.items():
    new_titles[info["new_dir"]] = old_titles[old_dir]

# 3. Create category directories and move files
for cat_name in categories.keys():
    cat_path = os.path.join(articles_dir, cat_name)
    os.makedirs(cat_path, exist_ok=True)

for old_dir, info in mapping.items():
    old_path = os.path.join(articles_dir, old_dir)
    new_path = os.path.join(articles_dir, info["cat"], info["new_dir"])
    
    if os.path.exists(old_path):
        # Move directory
        if os.path.exists(new_path):
            shutil.rmtree(new_path)
        shutil.move(old_path, new_path)

# 4. Process HTML files in the new structure
for cat_name in categories.keys():
    cat_path = os.path.join(articles_dir, cat_name)
    dirs = sorted([d for d in os.listdir(cat_path) if d.startswith("article-")])
    
    for i, d in enumerate(dirs):
        idx_path = os.path.join(cat_path, d, "index.html")
        if not os.path.exists(idx_path):
            continue
            
        with open(idx_path, "r", encoding="utf-8") as f:
            content = f.read()
            
        file_id = d.replace("article-", "") # e.g. "01-01"
        clean_t = new_titles[d]
        new_full_title = f"FILE {file_id}: {clean_t}"
        
        # Replace title block
        content = re.sub(r'{%\s*set title = "(.*?)"\s*%}', f'{{% set title = "{new_full_title}" %}}', content)
        
        # Replace h1 block
        content = re.sub(r'<h1>(.*?)</h1>', f'<h1>{new_full_title}</h1>', content)
        
        # Determine previous and next links
        prev_dir = dirs[i-1] if i > 0 else None
        next_dir = dirs[i+1] if i < len(dirs) - 1 else None
        
        # Replace bottom navigation block (heuristic: div containing '次の記事' or 'カテゴリ一覧' or 'ポータルへ')
        # Since layouts vary, we will replace the entire <div style="margin-top: 3rem; text-align: center;">...</div> at the end.
        
        # Find the last matching nav div at the bottom
        nav_pattern = r'<div style="margin-top: 3rem; text-align: center;">\s*<a href=.*?>.*?</a>\s*</div>'
        
        new_nav = '<div style="margin-top: 3rem; text-align: center; display: flex; justify-content: center; gap: 1rem; flex-wrap: wrap;">\n'
        
        if prev_dir:
            prev_title = new_titles[prev_dir]
            new_nav += f'    <a href="/articles/{cat_name}/{prev_dir}/" class="glass" style="padding: 1rem 2.5rem; color: var(--accent-gold); text-decoration: none; border: 1px solid var(--accent-gold); font-weight: bold; transition: all 0.3s;">&larr; 前へ：{prev_title}</a>\n'
        
        new_nav += f'    <a href="/#category" class="glass" style="padding: 1rem 2.5rem; color: #fff; text-decoration: none; border: 1px solid rgba(255,255,255,0.2); font-weight: bold; transition: all 0.3s;">カテゴリ一覧へ戻る</a>\n'

        if next_dir:
            next_title = new_titles[next_dir]
            new_nav += f'    <a href="/articles/{cat_name}/{next_dir}/" class="glass" style="padding: 1rem 2.5rem; color: var(--accent-gold); text-decoration: none; border: 1px solid var(--accent-gold); font-weight: bold; transition: all 0.3s;">次へ：{next_title} &rarr;</a>\n'
            
        new_nav += '</div>'
        
        if re.search(nav_pattern, content, re.DOTALL):
            content = re.sub(nav_pattern, new_nav, content, flags=re.DOTALL)
        else:
            # If not found, insert before </article>
            content = re.sub(r'</article>', f'{new_nav}\n</article>', content)
            
        # Also clean up any old portal links at the top
        content = re.sub(r'<nav style="margin-bottom: 2rem;">\s*<a href="/".*?>.*?</a>\s*</nav>', 
                         '<nav style="margin-bottom: 2rem;">\n        <a href="/" style="color: var(--accent-gold); text-decoration: none;">&larr; ポータルへ戻る</a>\n    </nav>', content)

        with open(idx_path, "w", encoding="utf-8") as f:
            f.write(content)

print("Migration completed.")

File: scratch/migrate_articles.js
const fs = require('fs');
const path = require('path');

const articlesDir = path.join('C:', 'Users', 'ntana', 'Documents', 'Antigravity_docs', 'digital-literacy-ip', 'src', 'articles');

const categories = {
    "01_p2p": ["01", "02", "03", "04", "05", "06", "07", "08", "09", "10", "20", "21", "22", "26", "27", "29", "30"],
    "02_apps": ["31", "32", "33", "34", "35", "36"],
    "03_clips": ["11", "12", "17", "19", "25"],
    "04_web": ["13", "14", "15", "16", "18", "28"]
};

function cleanTitle(title) {
    return title.replace(/^(?:FILE\s*\d+:?|FILE\s*\d+-\d+:?)\s*/i, '').trim();
}

const mapping = {};
const categoryArticles = {};
const newTitles = {};

// Build mapping
for (const [catName, oldNums] of Object.entries(categories)) {
    const catNum = catName.split('_')[0];
    categoryArticles[catName] = [];
    
    oldNums.forEach((oldNum, i) => {
        const newNum = String(i + 1).padStart(2, '0');
        const oldDir = `article-${oldNum}`;
        const newDir = `article-${catNum}-${newNum}`;
        mapping[oldDir] = { cat: catName, newDir, fileId: `${catNum}-${newNum}` };
        categoryArticles[catName].push(newDir);
    });
}

// Extract old titles
const oldTitles = {};
for (const oldDir of Object.keys(mapping)) {
    const idxPath = path.join(articlesDir, oldDir, 'index.html');
    if (fs.existsSync(idxPath)) {
        const content = fs.readFileSync(idxPath, 'utf8');
        const match = content.match(/<h1>(.*?)<\/h1>/);
        oldTitles[oldDir] = match ? cleanTitle(match[1]) : 'No Title';
    } else {
        oldTitles[oldDir] = 'No Title';
    }
}

for (const [oldDir, info] of Object.entries(mapping)) {
    newTitles[info.newDir] = oldTitles[oldDir];
}

// Create directories and move files
for (const catName of Object.keys(categories)) {
    const catPath = path.join(articlesDir, catName);
    if (!fs.existsSync(catPath)) fs.mkdirSync(catPath, { recursive: true });
}

for (const [oldDir, info] of Object.entries(mapping)) {
    const oldPath = path.join(articlesDir, oldDir);
    const newPath = path.join(articlesDir, info.cat, info.newDir);
    
    if (fs.existsSync(oldPath)) {
        if (fs.existsSync(newPath)) fs.rmSync(newPath, { recursive: true, force: true });
        fs.renameSync(oldPath, newPath);
    }
}

// Process HTML files
for (const catName of Object.keys(categories)) {
    const catPath = path.join(articlesDir, catName);
    const dirs = fs.readdirSync(catPath).filter(d => d.startsWith('article-')).sort();
    
    dirs.forEach((d, i) => {
        const idxPath = path.join(catPath, d, 'index.html');
        if (!fs.existsSync(idxPath)) return;
        
        let content = fs.readFileSync(idxPath, 'utf8');
        const fileId = d.replace('article-', '');
        const cleanT = newTitles[d];
        const newFullTitle = `FILE ${fileId}: ${cleanT}`;
        
        // Replace title block
        content = content.replace(/{%\s*set title = "(.*?)"\s*%}/g, `{% set title = "${newFullTitle}" %}`);
        
        // Replace h1 block
        content = content.replace(/<h1>(.*?)<\/h1>/g, `<h1>${newFullTitle}</h1>`);
        
        // Navigation links
        const prevDir = i > 0 ? dirs[i-1] : null;
        const nextDir = i < dirs.length - 1 ? dirs[i+1] : null;
        
        let newNav = '<div style="margin-top: 3rem; text-align: center; display: flex; justify-content: center; gap: 1rem; flex-wrap: wrap;">\n';
        
        if (prevDir) {
            newNav += `    <a href="/articles/${catName}/${prevDir}/" class="glass" style="padding: 1rem 2.5rem; color: var(--accent-gold); text-decoration: none; border: 1px solid var(--accent-gold); font-weight: bold; transition: all 0.3s;">&larr; 前へ：${newTitles[prevDir]}</a>\n`;
        }
        
        newNav += `    <a href="/" class="glass" style="padding: 1rem 2.5rem; color: #fff; text-decoration: none; border: 1px solid rgba(255,255,255,0.2); font-weight: bold; transition: all 0.3s;">ポータルへ戻る</a>\n`;

        if (nextDir) {
            newNav += `    <a href="/articles/${catName}/${nextDir}/" class="glass" style="padding: 1rem 2.5rem; color: var(--accent-gold); text-decoration: none; border: 1px solid var(--accent-gold); font-weight: bold; transition: all 0.3s;">次へ：${newTitles[nextDir]} &rarr;</a>\n`;
        }
        
        newNav += '</div>';
        
        const navPattern = /<div style="margin-top: 3rem; text-align: center;">\s*<a href=.*?<\/a>\s*<\/div>/s;
        if (navPattern.test(content)) {
            content = content.replace(navPattern, newNav);
        } else {
            content = content.replace(/<\/article>/, `${newNav}\n</article>`);
        }
        
        // Clean portal links
        content = content.replace(/<nav style="margin-bottom: 2rem;">\s*<a href="\/".*?>.*?<\/a>\s*<\/nav>/g, 
            '<nav style="margin-bottom: 2rem;">\n        <a href="/" style="color: var(--accent-gold); text-decoration: none;">&larr; ポータルへ戻る</a>\n    </nav>');

        fs.writeFileSync(idxPath, content, 'utf8');
    });
}

// Generate the mapping configuration for update_index.js to use
const indexMappingConfig = `module.exports = ${JSON.stringify(mapping, null, 4)};`;
fs.writeFileSync(path.join('C:', 'Users', 'ntana', 'Documents', 'Antigravity_docs', 'digital-literacy-ip', 'scratch', 'mapping.js'), indexMappingConfig, 'utf8');

console.log("Migration completed.");
